go/gildedrose/updatestrategy/strategyresolver.go

```diff
@@ -0,0 +1,33 @@
+package updatestrategy
+
+// enums to define the item types
+const (
+	_agedBrie  string = "Aged Brie"
+	_backStage string = "Backstage passes to a TAFKAL80ETC concert"
+	_sulfuras  string = "Sulfuras, Hand of Ragnaros"
+	_conjured  string = "Conjured Mana Cake"
+)
+
+// Only this interface and the methods are exposed outside 'updatestrategy' package, not the specific implementations
```

Comment:
nice!

Reply:
> Very cool, this pluggable `interface` is a nice approach. My Go experience is limited so I expected something like a different `struct` for each item type and then reflect on the type of the struct at run time to call the right update quality method or something like that 😃

Yes.. if you remember, I was trying to take an approach similar to what you are saying initially, except may be for the `reflect` part. Later I decided not to take that path because -
a) a specific kind of update(quality/sellin) is not something that happens "by" an item, but something that happens "to" the item. Which means, as long as there is an update, the `Item` struct itself shouldn't be concerned about `how` it happens. Makes it closed for unnecessary changes later on.
b) Say we want to implement a new update strategy later on, in the way I did it we could just plug in another strategy altogether, which makes the code extensible. Also, An update logic shouldn't be tightly coupled with a particular type. Ideally, I don't think a particular strategy should even be named like `agedBrieUpdateStrategy` but instead `incrementalUpdateStrategy`, `decrementalUpdateStrategy` or something else. But in our case the logic seemed coupled, and there was no evidence that the same logic may be used elsewhere, so I left it be. But a refactoring is possible given enough cases of reuse.

As of the use of reflection, I don't normally prefer it because there is a runtime performance overhead, especially when used for structs or while looping through an unlimited number of variables.

Reply:
@cyprusad